cmd/mt-whisper-importer-reader/main.go

```diff
@@ -248,123 +205,113 @@ func shortAggMethodString(aggMethod whisper.AggregationMethod) string {
 	}
 }
 
-func getMetric(w *whisper.Whisper, file string) (*archive.Metric, error) {
+func getMetrics(w *whisper.Whisper, file, name string) (archive.Metric, error) {
+	var res archive.Metric
 	if len(w.Header.Archives) == 0 {
-		return nil, errors.New(fmt.Sprintf("ERROR: Whisper file contains no archives: %q", file))
+		return res, fmt.Errorf("Whisper file contains no archives: %q", file)
 	}
 
-	archives := make([]archive.Archive, 0, len(w.Header.Archives))
-	var chunkSpan uint32
-	var rowKey string
-	var aggMethodStr string
-	name := getMetricName(file)
+	var archives []archive.Archive
+
+	method := shortAggMethodString(w.Header.Metadata.AggregationMethod)
+	if method == "" {
+		return res, fmt.Errorf(
+			"Aggregation method in file %s not allowed: %d(%s)\n",
+			file,
+			w.Header.Metadata.AggregationMethod,
+			method,
+		)
+	}
 
 	// md gets generated from the first archive in the whisper file
 	md := getMetricData(name, int(w.Header.Archives[0].SecondsPerPoint))
+	_, schema := schemas.Match(md.Name, 0)
 
-	for archiveIdx, archiveInfo := range w.Header.Archives {
-		if archiveIdx > 0 {
-
-			// on the first aggregation we determin the aggregation method as string
-			if aggMethodStr == "" {
-				aggMethodStr = shortAggMethodString(w.Header.Metadata.AggregationMethod)
-
-				if aggMethodStr == "" {
-					return nil, errors.New(fmt.Sprintf(
-						"ERROR: Aggregation method in file %s not allowed: %d(%s)\n",
-						file,
-						w.Header.Metadata.AggregationMethod,
-						aggMethodStr,
-					))
-				}
-			}
-
-			rowKey = api.AggMetricKey(
-				md.Id,
-				aggMethodStr,
-				archiveInfo.SecondsPerPoint,
-			)
-		} else {
-			rowKey = md.Id
-		}
-
-		// only read archive if archiveIdx is in readArchives
-		if _, ok := readArchives[archiveIdx]; !ok && len(readArchives) > 0 {
-			continue
-		}
-
-		encodedChunks := make([]chunk.IterGen, 0)
-
-		if len(chunkSpans)-1 < archiveIdx {
-			// if we have more archives than chunk spans are specified, we simply use the last one
-			chunkSpan = chunkSpans[len(chunkSpans)-1]
-		} else {
-			chunkSpan = chunkSpans[archiveIdx]
-		}
-
-		points, err := w.DumpArchive(archiveIdx)
+	points := make(map[int][]whisper.Point)
+	for i := range w.Header.Archives {
+		p, err := w.DumpArchive(i)
 		if err != nil {
-			return nil, errors.New(fmt.Sprintf("ERROR: Failed to read archive %d in %q, skipping: %q", archiveIdx, file, err))
+			return res, fmt.Errorf("Failed to dump archive %d from whisper file %s", i, file)
 		}
+		points[i] = p
+	}
 
-		var point whisper.Point
-		var t0, prevT0 uint32
-		var c *chunk.Chunk
-
-		for _, point = range sortPoints(points) {
-			// this shouldn't happen, but if it would we better catch it here because Metrictank wouldn't handle it well:
-			// https://github.com/raintank/metrictank/blob/f1868cccfb92fc82cd853914af958f6d187c5f74/mdata/aggmetric.go#L378
-			if point.Timestamp == 0 {
-				continue
+	conversion := newConversion(w.Header.Archives, points, method)
```

Comment:
this looks like it will only generate what whisper has.
so let's say for MT/HM we want avg,sum,cnt,min,max and whisper has sum, then we would only generate sum, whereas we could also generate cnt so that avg and cnt will work. I think?

Reply:
Yes, it only generates what Whisper has. 
We could add some number into `cnt` and make `avg` display something, yes. But since the whisper data is aggregated by `sum`, we can't reconstruct what would have been the `avg` value if whisper would have aggregated with `avg`, so the resulting data would be wrong, no?

Reply:
I was thinking that cnt is always trivially obtainable as rollup-interval / raw-interval, but i just realized that doesn't cut it for sparse data (e.g. when input data had a certain amount of nulls in it) so you're right that cnt (and hence) avg could be wrong.
but it's similar imho to the fake-avg case, where given an avg value, the cnt and sum values that we come up with will also be wrong if input data was sparse. so would it not make sense to be consistent across these cases (e.g. do them, with the known caveat of inaccuracy in case of sparse data) ?

Reply:
i just realized that you're right and with the "new way" of doing the fake `avg` we can easily do the same with the `cnt` with aggregation method is `sum`.
Why do you think that sparse data is going to be a problem? It will just have the same gaps in the resulting metrics, like shown in this test: https://github.com/raintank/metrictank/blob/importer_agg_conversion_plan/cmd/mt-whisper-importer-reader/conversion_test.go#L127

Reply:
let's say raw data was at 5s resolution:
```
ts val
0 5
10 5
15 null
20 5
25 10
30 5
35 5
40 5
45 5
50 5
55 10
```

whisper rollup with sum at 30s - the nulls don't contribute to the sum.
```
ts val
0 25
30 35
```

then imagine a few months later, we still have the rollup data, but not the raw data. and we want to generate cnt data so that avg works. the problem is we don't know how many values went into the sums, we could pretend the number is 6. so that means if you were to query for avg you would get values `25/6=4.167` and `35/6=5.833`. the first value is inaccurate.
but like i said, we can make this an explicit caveat by documenting it, since it's very similar to the current fakeavg possibly inaccuracies.

Reply:
ah, right. 
btw i'm now working on another commit where i basically reconstruct the fake avg data in every case where the source data has been aggregated by `sum`, that's going to include quite a few logic changes so maybe it's better if you hold on reviewing more right now.

Reply:
@Dieterbe the latest commit generates the correct `cnt` & `sum` values whenever a fake aggregation has to be generated based on an `avg` or `sum` archive. "correct" here means that the `cnt` is based on the given schema, as we discussed in the channel.